materialize-rockset/transactor.go

```diff
@@ -57,14 +58,48 @@ type transactor struct {
 	bindings []*binding
 }
 
+// awaitAllIntegrationIngestCompletions will block until all the Rockset collections named in the bindings
+// are in `READY` status and have completed any pending bulk ingestions. Specifically, this waits until the
+// number of objects in the bucket (as reported by rockset) and the number of successfully imported objects
+// is the same.
+func (t *transactor) awaitAllIntegrationIngestCompletions(ctx context.Context) error {
+	var waitGroup sync.WaitGroup
+	var errors = make([]error, len(t.bindings))
+	for i, binding := range t.bindings {
+		if binding.res.InitializeFromS3 != nil {
+			waitGroup.Add(1)
+			go func(errIndex int) {
+				errors[errIndex] = awaitCollectionReady(
+					ctx,
+					t.client,
+					binding.res.Workspace,
+					binding.res.Collection,
+					binding.res.InitializeFromS3.Integration,
+				)
+				waitGroup.Done()
+			}(i)
+		}
+	}
+	waitGroup.Wait()
+	for i, err := range errors {
+		if err != nil {
+			return fmt.Errorf("awaiting bulk ingestion completion for rockset collection '%s': %w",
+				t.bindings[i].rocksetCollection(), err)
```

Comment:
Do we specifically want to wait for all the collections to finish/error before reporting errors? I think you can use https://pkg.go.dev/golang.org/x/sync/errgroup to help here. It will handle propagating the first error from a routine up to the caller and cancelling the others.

Reply:
Oh neat, I didn't know about `errgroup`, but it's exactly what I wanted here. Thanks for the suggestion.